### PR TITLE
Fast randomness

### LINKED ISSUE #248: Allow all drand rounds
- [ ] computation of round gives the next drand round (not the next modulo 10)
- [ ] incentivise for rounds that are requested in the gateway
- [ ] Make the drand bots listen to what rounds are requested
- [ ] keep the current modulo 10 round incentive

### PR DESCRIPTION
**Pull request by webmaster128**
Closes #248

### CODE REVIEW

**Review by kaisbaccour — APPROVED** [review]
Just reviewed this looks super amazing. Well done.
One suggestion maybe for the future, maybe for now. How about removing the INCENTIVISED_ROUNDS entries when the slowest incentivised bot submits the randomness. 
So first bot submits then second then third .... the 6th bot removes the KV entry.

File: contracts/nois-drand/src/state.rs [hunk 1]
```diff
@@ -75,6 +93,12 @@ pub const SUBMISSIONS: Map<(u64, &Addr), StoredSubmission> = Map::new("submissio
 /// The number of submissions done for each round
 pub const SUBMISSIONS_COUNT: Map<u64, u16> = Map::new("counts");
 
+/// Dummy value. Don't rely on the value but just check existence.
+pub const INCENTIVIZED_BY_GATEWAY_MARKER: u8 = 1;
+
+/// A map from round number to marker value
+pub const INCENTIVIZED_BY_GATEWAY: Map<u64, u8> = Map::new(TopKey::Incentivised.as_str());
```
Comment (kaisbaccour) [inline contracts/nois-drand/src/state.rs:100]: why not Map<u64, ()> ?

**Review by kaisbaccour — APPROVED** [review]

File: contracts/nois-drand/src/state.rs (hunk 1 above)
Comment (webmaster128) [reply contracts/nois-drand/src/state.rs:100]: `()` is serialized to `null` in JSON, which is 4 bytes instead of 1

Comment (kaisbaccour) [reply contracts/nois-drand/src/state.rs:100]: haha of course. thanks.

**Comment by webmaster128** [comment]
> One suggestion maybe for the future, maybe for now. How about removing the INCENTIVISED_ROUNDS entries when the slowest incentivised bot submits the randomness.
> So first bot submits then second then third .... the 6th bot removes the KV entry.

Good idea! I started adding it but I think it's better to do it as part of https://github.com/noislabs/nois-contracts/issues/154. Then we can do all cleanups for a round together and bump `config.min_round` as a way to finalize those rounds.

**Comment by kaisbaccour** [comment]
> > One suggestion maybe for the future, maybe for now. How about removing the INCENTIVISED_ROUNDS entries when the slowest incentivised bot submits the randomness.
> > So first bot submits then second then third .... the 6th bot removes the KV entry.
> 
> Good idea! I started adding it but I think it's better to do it as part of #154. Then we can do all cleanups for a round together and bump `config.min_round` as a way to finalize those rounds.

This is not only for cleanup but also to make sure new bots don't try to submit to that requested round.  or is that already the case because after the 6th it becomes not incentivised ?

**Comment by webmaster128** [comment]
> or is that already the case because after the 6th it becomes not incentivised ?

This part does not change:

```
    // Check if the bot is fast enough to get an incentive
    if submissions_count < NUMBER_OF_INCENTIVES_PER_ROUND {
        reward_points += INCENTIVE_POINTS_FOR_FAST_BOT;
    }
```

So for both mod10 rounds and job rounds we have up to 6 incentivised submissions

**Comment by kaisbaccour** [comment]
> > or is that already the case because after the 6th it becomes not incentivised ?
> 
> This part does not change:
> 
> ```
>     // Check if the bot is fast enough to get an incentive
>     if submissions_count < NUMBER_OF_INCENTIVES_PER_ROUND {
>         reward_points += INCENTIVE_POINTS_FOR_FAST_BOT;
>     }
> ```
> 
> So for both mod10 rounds and job rounds we have up to 6 incentivised submissions

Okay clear, thanks

**Pull request merged**